README.md
# Fedimg

Service to automatically upload built Fedora images to internal and external
cloud providers.

## License

Fedimg is licensed under AGPLv3 or newer. See the `LICENSE` file for more
information.

## Documentation

Official Fedimg documentation can be [found at
RTFD](https://fedimg.readthedocs.org) or in `docs/`.

## Triggering jobs

Fedimg is designed to perform automatically when it sees the appropriate
fedmsg. However, sometimes, it's useful to be able to quickly trigger
a job manually. If Fedimg is installed properly, the `bin/trigger_upload.py`
script can be used for this purpose:

```
./bin/trigger_upload.py SOME_RAWXZ_URL
```

This script simply skips the part where Fedimg listens for the fedmsg, and
allows you to directly provide a URL to a raw.xz image file that you'd like
uploaded. Otherwise, Fedimg performs the same as during automatic operation.

## Providers

We hope to simultaneously upload our cloud images to a variety of internal and
external spaces. Currently, the code supports Amazon EC2. Work has begun
toward supporting Rackspace, GCE, and HP. We're currently waiting on some
legal developments to determine what sort of account and access we'll have
to these providers.

## Contributors

* David Gay <dgay@redhat.com>

* Ralph Bean <rbean@redhat.com>


CHANGELOG.md

## 0.6

**General**
-   use a single threadpool for all uploads to avoid leaking threads
-   prevent major IndexError when checking Koji tasks that don't have raw.xz
    outputs
-   increase number of fedmsg endpoints

**EC2Service**
-   use larger and more powerful instance types for utility and test instances
-   typofix when naming PV images

**Docs**
-   add some basic contributor docs

## 0.5

**EC2Service**
-   use 7 GB volume size rather than 3 GB for now, since atomic images come out
    to be 6.1 GB
-   implement gp2 volume type uploads
-   image name now includes volume type
-   simplify consumer filter code, eliminating 32 bit stuff for now
-   add build name, virtualization type, and volume type to 'extra'
    dict in fedmsgs

**Tests**
-   fix up consumer test code
-   add additional consumer tests to test build filter code

**Docs**
-   add info about volume size configuration
-   "tested on F21"
-   improve index page
-   bring installation info up-to-date

**Misc**
-   commit atomic test script, to go with base test script
-   reduce description in setup.py

## 0.4

**EC2Service**
-   fix alternate destinations not being set properly during image copy
-   split util and test AMIs into dedicated lists
-   allow for URL redirection while curling raw.xz image
-   simplified registration AKI selection process
-   major refactoring to allow for future expansion into many different
    types of AMIs
-   uploads are now multithreaded
-   volume size options added to config options
-   better logging
-   close a dangling SSH connection (thanks, threebean!)
-   fix bug that caused only the first two AMIs to be made public

**Tests**
-   fix broken consumer test
-   committed `uploadtest.py` for doing EC2Service test runs during
    development

**Docs**
-   update messaging docs
-   add table of AMI types to EC2Service docs
-   add AMI config format info

**Misc**
-   removed extraneous EC2Service-specific stuff from other service files
-   better commenting

## 0.3.2

-   use fedmsg logging utilities
-   convert old print statements to logging

## 0.3.1

-   cycle through and make copied AMIs public after uploads complete
-   register AMI with description containing build name of source image file
-   report AMI Ids when emitting related fedmsgs
-   make sure all AMIs have a matching numerical extension across regions
-   clean up a little EC2Service code
-   typofixes, etc

## 0.3

-   add utility function to get virtualization type for EC2 AMI registration
-   make AMIs public after being tested and cpied
-   tweaks to layout of config file
-   only use 64 bit EBS utility instances
-   remove hardcoded username
-   rename some variables to be clearer
-   add clean_up_on_failure and delete_image_on_failure config options
-   improve exception handling
-   make sure snapshot is deleted on cleanup if no image has been registered
-   add some starter tests
-   move around some processes to more appropriate files
-   don't attempt to provide an AKI when registering an image as HVM
-   fix root device name for PV vs. HVM
-   serious PEP 8 fixes
-   fix up duplicate image name prevention code
-   various typofixes and code cleanup

## 0.2.6

-   use proper buildroot macro in spec file
-   preserve file timestamps when copying in spec file
-   do not make library files executable, and don't give them shebangs
-   add license headers to all Python files 

## 0.2.5

-   Remove coding from fedmsg.d/fedimg.py since it seems to make it executable
-   make init file executable in spec install section, as well

## 0.2.4

-   Shorten spec summary and remove trailing dot
-   Add shebang and coding to top of fedimg init file
-   Remove shebang from fedmsg.d/fedimg.py
-   Make all necessary fedimg files executable in spec install section

## 0.2.3

-   Better IAM profile name example in the config
-   Addition to docs: section about setting up the config file
-   Fix strange saces and add missing comma to setup.py classifiers section

## 0.2.2

-   Include .pyc and .pyo files for consumer in /etc/fedmsg.d/
-   Add missing comma

## 0.2.1

-   Fix `packages` argument in setup.py to take `find_packages()`

## 0.2.0

-   Initial RPM release to Fedora
-   setup.py improvements
-   Config file is now read from /etc/fedimg.cfg
-   PEP 8 fixes

## 0.1.0

-   Initial PyPI release



fedimg/consumers.py
# This file is part of fedimg.
# Copyright (C) 2014 Red Hat, Inc.
#
# fedimg is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# fedimg is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public
# License along with fedimg; if not, see http://www.gnu.org/licenses,
# or write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#
# Authors:  David Gay <dgay@redhat.com>
#

import logging
log = logging.getLogger("fedmsg")

import multiprocessing.pool

import fedmsg.consumers
import fedmsg.encoding
import koji

import fedimg.uploader
from fedimg.util import get_rawxz_url


class KojiConsumer(fedmsg.consumers.FedmsgConsumer):
    """ Listens for image Koji task completion and sends image files
        produced by the child createImage tasks to the uploader. """
    # To my knowledge, all *image* builds appear under this
    # exact topic, along with scratch builds.
    topic = 'org.fedoraproject.prod.buildsys.task.state.change'
    config_key = 'kojiconsumer'

    def __init__(self, *args, **kwargs):
        # Patch libcloud to give it support for the ec2 eu-central-1 region
        import fedimg.haxx
        fedimg.haxx.monkeypatch_libcloud()

        super(KojiConsumer, self).__init__(*args, **kwargs)

        # threadpool for upload jobs
        self.upload_pool = multiprocessing.pool.ThreadPool(processes=4)

        log.info("Super happy fedimg ready and reporting for duty.")

    def _get_upload_urls(self, builds):
        """ Takes a list of koji createImage task IDs and returns a list of
        URLs to .raw.xz image files that should be uploaded. """

        for build in builds:
            log.info('Got Koji build {0}'.format(build))

        # Create a Koji connection to the Fedora Koji instance
        koji_session = koji.ClientSession(fedimg.KOJI_SERVER)

        rawxz_files = []  # list of full URLs of files

        # Get all of the .raw.xz URLs for the builds
        if len(builds) == 1:
            task_result = koji_session.getTaskResult(builds[0])
            url = get_rawxz_url(task_result)
            if url:
                rawxz_files.append(url)
        elif len(builds) >= 2:
            koji_session.multicall = True
            for build in builds:
                koji_session.getTaskResult(build)
            results = koji_session.multiCall()
            for result in results:
                if not result: continue
                url = get_rawxz_url(result[0])
                if url:
                    rawxz_files.append(url)

        # We only want to upload:
        # 64 bit: base, atomic, bigdata
        # Not uploading 32 bit, vagrant, experimental, or other images.
        upload_files = []  # files that will actually be uploaded
        for url in rawxz_files:
            u = url.lower()
            if u.find('x86_64') > -1 and u.find('vagrant') == -1:
                if (u.find('fedora-cloud-base') > -1
                        or u.find('fedora-cloud-atomic') > -1
                        or u.find('fedora-cloud-bigdata') > -1):
                    upload_files.append(url)
                    log.info('Image {0} will be uploaded'.format(url))

        return upload_files

    def consume(self, msg):
        """ This is called when we receive a message matching the topic. """

        builds = list()  # These will be the Koji build IDs to upload, if any.

        msg_info = msg["body"]["msg"]["info"]

        log.info('Received %r %r' % (msg['topic'], msg['body']['msg_id']))

        # If the build method is "image", we check to see if the child
        # task's method is "createImage".
        if msg_info["method"] == "image":
            if isinstance(msg_info["children"], list):
                for child in msg_info["children"]:
                    if child["method"] == "createImage":
                        # We only care about the image if the build
                        # completed successfully (with state code 2).
                        if child["state"] == 2:
                            builds.append(child["id"])

        if len(builds) > 0:
            fedimg.uploader.upload(self.upload_pool,
                                   self._get_upload_urls(builds))
